nova_vm/src/ecmascript/abstract_operations/testing_and_comparison.rs

```diff
@@ -53,37 +56,61 @@ pub(crate) fn is_array(_agent: &Agent, argument: Value) -> JsResult<bool> {
 /// The abstract operation IsCallable takes argument argument (an ECMAScript
 /// language value) and returns a Boolean. It determines if argument is a
 /// callable function with a [[Call]] internal method.
-pub(crate) fn is_callable(argument: impl IntoValue) -> bool {
+///
+/// > #### Note
+/// > Nova breaks with the specification to narrow the types automatically, and
+/// > returns an `Option<Function>`. Eventually this should become
+/// > `Option<Callable>` once callable proxies are supported.
+pub(crate) fn is_callable(argument: impl TryInto<Function>) -> Option<Function> {
     // 1. If argument is not an Object, return false.
     // 2. If argument has a [[Call]] internal method, return true.
     // 3. Return false.
-    matches!(
-        argument.into_value(),
-        Value::BoundFunction(_)
-            | Value::BuiltinFunction(_)
-            | Value::ECMAScriptFunction(_)
-            | Value::BuiltinPromiseResolvingFunction(_)
-    )
+    argument.try_into().ok()
 }
 
-pub(crate) fn is_constructor(agent: &mut Agent, constructor: Value) -> bool {
-    // If argument is not an Object, return false.
+/// ### [7.2.4 IsConstructor ( argument )](https://tc39.es/ecma262/#sec-isconstructor)
+///
+/// The abstract operation IsConstructor takes argument argument (an ECMAScript
+/// language value) and returns a Boolean. It determines if argument is a
+/// function object with a [[Construct]] internal method.
+///
+/// > #### Note
+/// > Nova breaks with the specification to narrow the types automatically, and
+/// > returns an `Option<Function>`. Eventually this should become
+/// > `Option<Callable>` once callable proxies are supported.
+pub(crate) fn is_constructor(agent: &mut Agent, constructor: impl IntoValue) -> Option<Function> {
```

Comment:
The spec concept of callable matches one to one to our concept of `Function`, but that is not the same for constructors. Would it make sense to split this function into `Function::is_constructor` (that takes a `Function` and returns a boolean), and `to_constructor`, that takes a value (or `IntoValue`) and returns an `Option<Function>`?

Reply:
Yeah, that definitely sounds like a good idea, since it's essentially determining if a Function has a `[[Construct]]` internal method.

Reply:
I split the method so that the `is_constructor` now takes an `impl TryInto<Function>` like `is_callable`, and calls into `Function::is_constructor` which delegates to the different Function variants for the internal details of how constructorhood is defined.